capabilityquerier/pkg/capabilityquerier/capabilityquerier.go

```diff
@@ -246,80 +248,65 @@ func requestCapabilityStatementAndSmartOnFhir(ctx context.Context, fhirURL strin
-		}
-		httpResponseCode, tlsVersion, mimeTypeWorked, capResp, responseTime, err = requestWithMimeType(req, firstMIME, client)
-		if err != nil {
-			return err
-		}
-	} else if endptType == wellknown && len(message.MIMETypes) > 0 {
-		firstMIME = message.MIMETypes[0]
-		httpResponseCode, _, _, capResp, _, err = requestWithMimeType(req, firstMIME, client)
-		if err != nil {
-			return err
-		}
-	} else {
-		httpResponseCode, tlsVersion, mimeTypeWorked, capResp, responseTime, err = requestWithMimeType(req, fhir3PlusJSONMIMEType, client)
-		if err != nil {
+	if len(message.MIMETypes) == 1 {
+		// Only one MIME type saved
+		savedMIME := message.MIMETypes[0]
+		httpResponseCode, tlsVersion, mimeTypeWorked, capResp, responseTime, httpErr = requestWithMimeType(req, savedMIME, client)
+		if httpErr != nil && httpResponseCode != 0 {
 			return err
 		}
 	}
 
-	if endptType == metadata {
-		otherMime := fhir2LessJSONMIMEType
-		if httpResponseCode != http.StatusOK || !mimeTypeWorked {
-			// Try the other mime type and remove the mime type that was initially saved
-			// but no longer works
-			if len(message.MIMETypes) == 2 {
-				otherMimeIdx := (randomMimeIdx + 1) % 2
-				otherMime = message.MIMETypes[otherMimeIdx]
-				message.MIMETypes = []string{otherMime}
-			} else if len(message.MIMETypes) == 1 {
-				if message.MIMETypes[0] == otherMime {
-					otherMime = fhir3PlusJSONMIMEType
+	if len(message.MIMETypes) != 1 || httpResponseCode != http.StatusOK || !mimeTypeWorked {
+		if endptType == wellknown {
+			if len(message.MIMETypes) == 0 {
+				httpResponseCode, _, _, capResp, _, httpErr = requestWithMimeType(req, fhir3PlusJSONMIMEType, client)
+				if httpErr != nil && httpResponseCode != 0 {
+					return err
 				}
+			}
+		} else if endptType == metadata {
+			oldMIMEType := ""
+			if len(message.MIMETypes) == 1 {
+				oldMIMEType = message.MIMETypes[0]
+				message.MIMETypes = []string{}
+			} else if len(message.MIMETypes) > 1 {
 				message.MIMETypes = []string{}
 			}
-			// replace all values based on the other mime type if there were any issues with the first mime type request
-			httpResponseCode, tlsVersion, otherMimeWorked, capResp, responseTime, err = requestWithMimeType(req, otherMime, client)
-			if err != nil {
-				return err
+
+			if oldMIMEType != fhir2LessJSONMIMEType {
+				httpResponseCode, tlsVersion, mimeTypeWorked, capResp, responseTime, httpErr = requestWithMimeType(req, fhir2LessJSONMIMEType, client)
+				if httpErr != nil && httpResponseCode != 0 {
+					return err
+				}
+				triedMIMEType = fhir2LessJSONMIMEType
 			}
-		} else if len(message.MIMETypes) == 0 {
-			// only check fhir 2 mime type support if the first request worked and there were no
-			// mimeTypes saved in the database
-			_, _, otherMimeWorked, _, _, err = requestWithMimeType(req, otherMime, client)
-			if err != nil {
-				return err
+			if oldMIMEType != fhir3PlusJSONMIMEType && (!mimeTypeWorked || httpResponseCode != http.StatusOK) {
```

Comment:
I've been waffling back and forth on this, but I currently feel like we should actually check the fhir3PlusJSONMIMEType first, since more of our endpoints are becoming r4 endpoints. I also feel like it tells us more if we use that one first, cause then an r4 endpoint that has the fhir2Less mime type saved would mean that the fhir3plus one isn't working, and they're therefore non-compliant, whereas if we check the fhir2Less first and it works, all we know is that they support the fhir2less, we don't necessarily know that they support the fhir3plus as well. what do you think?

Reply:
That makes a lot of sense, I will change this!